x-pack/plugins/security/server/saved_objects/secure_saved_objects_client_wrapper.ts

```diff
@@ -618,20 +551,234 @@ export class SecureSavedObjectsClientWrapper implements SavedObjectsClientContra
+    // for, and 3. create audit records for any objects that will be returned to the user.
+    const requestedObjectsSet = objects.reduce(
+      (acc, { type, id }) => acc.add(`${type}:${id}`),
+      new Set<string>()
+    );
+    const traversedObjects = new Set<string>();
+    const filteredObjectsMap = new Map<string, SavedObjectReferenceWithContext>();
+    const getIsAuthorizedForInboundReference = (inbound: { type: string; id: string }) => {
+      const found = filteredObjectsMap.get(`${inbound.type}:${inbound.id}`);
+      return found && !found.isMissing; // If true, this object can be linked back to one of the requested objects
+    };
+    let loopCount = 0;
+    let objectsToProcess = [...response.objects];
+    while (++loopCount && objectsToProcess.length > 0) {
+      const obj = objectsToProcess.shift()!;
+      const { type, id, spaces, inboundReferences } = obj;
+      const objKey = `${type}:${id}`;
+      traversedObjects.add(objKey);
+      // Is the user authorized to access this object in all required space(s)?
+      const isAuthorizedForObject = isAuthorizedForObjectInAllSpaces(
+        type,
+        authAction,
+        typeActionMap,
+        [currentSpaceId]
+      );
+      // Redact the inbound references so we don't leak any info about other objects that the user is not authorized to access
+      const redactedInboundReferences = inboundReferences.filter((inbound) => {
+        if (inbound.type === type && inbound.id === id) {
+          // circular reference, don't redact it
+          return true;
+        }
+        return getIsAuthorizedForInboundReference(inbound);
+      });
+      // If the user is not authorized to access at least one inbound reference of this object, then we should omit this object.
+      const isAuthorizedForGraph =
+        requestedObjectsSet.has(objKey) || // If true, this is one of the requested objects, and we checked authorization above
+        redactedInboundReferences.some(getIsAuthorizedForInboundReference);
+
+      if (isAuthorizedForObject && isAuthorizedForGraph) {
+        if (spaces.length) {
+          // Don't generate audit records for "empty results" with zero spaces (requested object was a non-multi-namespace type or hidden type)
+          this.auditLogger.log(
+            savedObjectEvent({
+              action: SavedObjectAction.COLLECT_MULTINAMESPACE_REFERENCES,
+              savedObject: { type, id },
+            })
+          );
+        }
+        filteredObjectsMap.set(objKey, obj);
+      } else if (!isAuthorizedForObject && isAuthorizedForGraph) {
+        filteredObjectsMap.set(objKey, { ...obj, spaces: [], isMissing: true });
+      } else if (
+        isAuthorizedForObject &&
+        !isAuthorizedForGraph &&
+        inboundReferences.some((x) => !traversedObjects.has(`${x.type}:${x.id}`)) &&
+        loopCount <= response.objects.length // circuit-breaker to prevent infinite loops
```

Comment:
I'm not so sure I agree. It *shouldn't* happen anyway, but if you happen to have a case that would trigger this circuit-breaker (as demonstrated in the unit test): if Security is disabled the operation will succeed, but if Security is enabled the operation will fail with a new type of error. For any other object that the user did not request (a reference that was found but the user was not authorized to access it), we treat it as a "not found" result and just omit it. Why would we do anything different here?

Perhaps "circuit-breaker" is the wrong word to describe what I'm trying to accomplish here?

Reply:
I think it's dangerous to assume that we are in a valid state if we encounter a scenario that "shouldn't happen".

If we have a pointer to a reference that wasn't retrieved by the base client, then that is indicative of a bug, and I don't think we should treat that as a successful outcome.

> Perhaps "circuit-breaker" is the wrong word to describe what I'm trying to accomplish here?

Yeah, that could be it. Instead of counting loop iterations, what if we instead ensure that the object has at least 1 _valid_ inbound reference before sending it to the back for future processing? Would that accomplish what we're looking for?

```diff
diff --git a/x-pack/plugins/security/server/saved_objects/secure_saved_objects_client_wrapper.ts b/x-pack/plugins/security/server/saved_objects/secure_saved_objects_client_wrapper.ts
index b096db77b59..9eabd0dd0c0 100644
--- a/x-pack/plugins/security/server/saved_objects/secure_saved_objects_client_wrapper.ts
+++ b/x-pack/plugins/security/server/saved_objects/secure_saved_objects_client_wrapper.ts
@@ -613,6 +613,10 @@ export class SecureSavedObjectsClientWrapper implements SavedObjectsClientContra
       (acc, { type, id }) => acc.add(`${type}:${id}`),
       new Set<string>()
     );
+    const retrievedObjectsSet = response.objects.reduce(
+      (acc, { type, id }) => acc.add(`${type}:${id}`),
+      new Set<string>()
+    );
     const traversedObjects = new Set<string>();
     const filteredObjectsMap = new Map<string, SavedObjectReferenceWithContext>();
     const getIsAuthorizedForInboundReference = (inbound: { type: string; id: string }) => {
@@ -659,14 +663,29 @@ export class SecureSavedObjectsClientWrapper implements SavedObjectsClientContra
         filteredObjectsMap.set(objKey, obj);
       } else if (!isAuthorizedForObject && isAuthorizedForGraph) {
         filteredObjectsMap.set(objKey, { ...obj, spaces: [], isMissing: true });
-      } else if (
-        isAuthorizedForObject &&
-        !isAuthorizedForGraph &&
-        inboundReferences.some((x) => !traversedObjects.has(`${x.type}:${x.id}`)) &&
-        loopCount <= response.objects.length // circuit-breaker to prevent infinite loops
-      ) {
-        // this object has inbound reference(s) that we haven't traversed yet; bump it to the back of the list
-        objectsToProcess = [...objectsToProcess, obj];
+      } else if (isAuthorizedForObject && !isAuthorizedForGraph) {
+        const hasUntraversedInboundReferences = inboundReferences.some(
+          (ref) =>
+            !traversedObjects.has(`${ref.type}:${ref.id}`) &&
+            retrievedObjectsSet.has(`${ref.type}:${ref.id}`)
+        );
+
+        if (hasUntraversedInboundReferences) {
+          // this object has inbound reference(s) that we haven't traversed yet; bump it to the back of the list
+          objectsToProcess = [...objectsToProcess, obj];
+        } else {
+          // There should never be a missing inbound reference.
+          // If there is, then something has gone terribly wrong.
+          const hasMissingInboundReference = inboundReferences.some(
+            (ref) =>
+              !traversedObjects.has(`${ref.type}:${ref.id}`) &&
+              !retrievedObjectsSet.has(`${ref.type}:${ref.id}`)
+          );
+
+          if (hasMissingInboundReference) {
+            throw new Error('This was not supposed to happen.');
+          }
+        }
       }
     }
 

 
```

Reply:
I guess part of my reasoning for "this shouldn't happen, but it could happen" is if another SOC wrapper is added that changes the functionality in this way (by completely omitting some object in the middle of the reference chain, intentionally or otherwise). But we aren't planning on doing that and I don't see why we would need to support it, so I guess it's a bit of a moot point anyway and it should be OK to throw an error in that case.

> Instead of counting loop iterations, what if we instead ensure that the object has at least 1 _valid_ inbound reference before sending it to the back for future processing? Would that accomplish what we're looking for?

Sure, it looks like it accomplishes basically the same thing but the logic is certainly easier to follow. Thanks for coming up with that, I'll add it in.